### DISCUSSION TITLE
Support -nostdlib

### DISCUSSION THREAD

**Discussion by theoparis** [scala-native/scala-native#3661] [Ideas]
This would allow true bare metal code to be created with clang. I tried specifying it in withCompileOptions but it resulted in javalib failing to compile. I could not find a way to disable the javalib/standard libraries completely.

Edit: It seems like there is also no way to access raw memory manually (using addresses without malloc). It also seems like there's no way to define a custom main entrypoint

**Comment by WojciechMazur** [comment]
Hey, thank you for bringing up this topic. To make it happend we should start with eliminating C++ dependency, but there are at least 2 blockers for that right now: 
 1. Current exception handling is based on C++ exceptions under the hood. We'd need to provide a runtime for personality functions first and modify the codegen. 
 2. We use embedded code of libunwind to provide stacktraces - this code is using c++. Even if we'd use an already published library we'll still need  
 
For both of them we currently need to provide C++ runtime. There is a goal to remove this dependency, but it had never priority big enough to spend precious resources on that.

There's also a long list of functions which are always rechable and linked to allow for safe code lowering, we'd need to handle this as well. 

However, in my opinion, Scala Native was never aimed to be a target for really low-level programming. The main purpose was to move existing high level language out of the JVM and it's limitations, while keeping similar code semantics and capabilities. In it's current state it's strongly coupled with both GC and runtime using large amount of C stdlib + POSIX/WindowsAPI to reimplement most of Java Standard Library.

**Reply by ekrich** [reply]
As pointed out above the C++ is a big issue to remove and not really a big plus for Scala Native as it is. In order to have the Scala Language we need to have at least the basic Java lib that is needed to support the runtime and basis of Scala that is dependent on the Java core. GC is also needed for the runtime because Scala is a JVM language and that is just part of the deal.

To have a small runtime could be a potential goal and might be workable as there could be a core / minimal Java lib. But as mentioned, even to get there is a tall order but not impossible. Very small embedded devices seems a non-goal but platforms that can run a minimal Scala Native image is possible. How small is the question.
